File: 이재진/무인도여행.py
from collections import deque
def solution(maps):
    answer = []
    visited = [[-1]*len(maps[0]) for _ in range(len(maps))]

    def bfs(i,j):
        dij = [[0,1], [1,0], [0,-1], [-1,0]]
        q = deque()
        q.append([i,j])
        ans = 0
        while q:
            ci, cj = q.popleft()
            ans += maps[ci][cj]
            for di,dj in dij:
                ni,nj = ci+di, cj+dj
                if maps[ni][nj] != "X" and not visited[ni][nj]:
                    q.append([ni,nj])
                    visited[ni][nj] = 1
        return ans
    
    dij = [[0,1], [1,0], [0,-1], [-1,0]]
    for i in range(len(maps)):
        for j in range(len(maps[i])):
            if maps[i][j] != "X" and visited[i][j] == -1:
                q = deque()
                q.append([i,j])
                ans = 0
                while q:
                    ci, cj = q.popleft()
                    visited[ci][cj] = 1
                    ans += int(maps[ci][cj])
                    for di,dj in dij:
                        ni,nj = ci+di, cj+dj
                        if 0 <= ni < len(maps) and 0 <= nj < len(maps[0]) and maps[ni][nj] != "X" and visited[ni][nj] == -1:
                            q.append([ni,nj])
                            visited[ni][nj] = 1
                answer.append(ans)
        answer.sort()
        if not answer:
            answer.append(-1)
    return answer

File: 정종길/프로그래머스/기타/무인도 여행.java
import java.util.*;

class Solution {
    static boolean[][] visited ;
    static int row;
    static int col;

    public List<Integer> solution(String[] maps) {
        List<Integer> answer = new ArrayList<>();
        row = maps.length;
        col = maps[0].length();
        visited = new boolean[row][col];

        for(int i = 0; i < row; i ++ ){
            for(int j = 0; j < col; j ++){
                if (maps[i].charAt(j) != 'X' && visited[i][j] == false){
                    int distance = bfs(maps[i].charAt(j), i, j, maps);
                    answer.add(distance);
                }
            }
        }

        if(answer.isEmpty()){
            answer.add(-1);
        } else {
            Collections.sort(answer);
        }
        return answer;
    }

    public int bfs(char num, int i, int j, String[] maps){
        Queue<Integer[]> queue = new LinkedList<>();
        int area = 0;
        int[] dx = {0, 1, 0, -1};
        int[] dy = {1, 0, -1, 0};
        queue.add(new Integer[]{Integer.parseInt(String.valueOf(num)), i, j});

        visited[i][j] = true;

        while(!queue.isEmpty()){
            Integer[] arr = queue.poll();
            area += arr[0];
            int x = arr[1];
            int y = arr[2];

            for(int k = 0; k < 4; k ++){
                int nx = x + dx[k];
                int ny = y + dy[k];
                if(0 <= nx && nx < row && 0 <= ny && ny < col){
                    if(visited[nx][ny] == false && maps[nx].charAt(ny) != 'X'){
                        visited[nx][ny] = true;
                        queue.add(new Integer[] { Integer.parseInt(String.valueOf(maps[nx].charAt(ny))), nx, ny });
                    }
                }
            }

        }
        return area;
    }

}